File: src/core/natural_sort.py
"""Natural sort key helper.

Splits a string into numeric and non-numeric chunks so that "KHW_10" sorts
after "KHW_2" (instead of lexicographically before it).

Spec: docs/specs/modules/gpx_multi_import.md (Issue #127).
"""
from __future__ import annotations

import re

_NUM_RE = re.compile(r"(\d+)")


def natural_sort_key(s: str) -> list:
    """Return a sort key that orders embedded integers numerically.

    Examples:
        sorted(["KHW_11", "KHW_00a", "KHW_10"], key=natural_sort_key)
            == ["KHW_00a", "KHW_10", "KHW_11"]
        sorted(["10.gpx", "2.gpx", "1.gpx"], key=natural_sort_key)
            == ["1.gpx", "2.gpx", "10.gpx"]

    Implementation:
        re.split with a capture group keeps numeric tokens as separate
        list elements (alternating with text). Even indices are text
        (lower-cased for case-insensitive ordering), odd indices are
        numeric tokens converted to int for numeric ordering.
    """
    parts = _NUM_RE.split(s)
    return [
        int(p) if i % 2 == 1 else p.lower()
        for i, p in enumerate(parts)
    ]

File: frontend/src/lib/utils/naturalSort.ts
/**
 * Natural sort helpers — order strings with embedded integers numerically
 * so "KHW_10" sorts AFTER "KHW_2" instead of lexicographically before it.
 *
 * Mirror of `src/core/natural_sort.py::natural_sort_key()` for the
 * SvelteKit-Frontend (Spec: docs/specs/modules/gpx_multi_import.md v1.1).
 */

/**
 * Build a sort key for a string by splitting it into alternating
 * non-digit and digit chunks. Digit chunks become integers; text chunks
 * are lower-cased so the sort is case-insensitive.
 *
 * Examples:
 *   naturalSortKey('KHW_10') -> ['khw_', 10, '']
 *   naturalSortKey('KHW_00a') -> ['khw_', 0, 'a']
 */
export function naturalSortKey(s: string): Array<string | number> {
	// /(\d+)/ with capturing group keeps numeric runs as separate parts.
	// Even indices (0, 2, ...) are non-digit text; odd indices are digit runs.
	const parts = s.split(/(\d+)/);
	return parts.map((part, i) =>
		i % 2 === 1 ? parseInt(part, 10) : part.toLowerCase()
	);
}

/**
 * Stable sort of `arr` by the natural-sort key derived from `key(item)`.
 * Returns a new array — does not mutate the input.
 */
export function naturalSort<T>(arr: T[], key: (item: T) => string): T[] {
	const decorated = arr.map((item, idx) => ({
		item,
		idx,
		sortKey: naturalSortKey(key(item))
	}));
	decorated.sort((a, b) => {
		const ka = a.sortKey;
		const kb = b.sortKey;
		const len = Math.max(ka.length, kb.length);
		for (let i = 0; i < len; i++) {
			const xa = ka[i];
			const xb = kb[i];
			if (xa === undefined && xb === undefined) continue;
			if (xa === undefined) return -1;
			if (xb === undefined) return 1;
			// Mixed string/number: compare via string repr to remain total-orderable.
			if (typeof xa !== typeof xb) {
				const sa = String(xa);
				const sb = String(xb);
				if (sa < sb) return -1;
				if (sa > sb) return 1;
				continue;
			}
			if (xa < xb) return -1;
			if (xa > xb) return 1;
		}
		return a.idx - b.idx; // stability for equal keys
	});
	return decorated.map((d) => d.item);
}
